### PR TITLE
fix: target_container incorrectly using main container port/protocol

### LINKED ISSUE #4041: target_container option not working as expected
Website Manifest
```
name: website
type: Load Balanced Web Service

http:
  path: '/'
  target_container: proxy

image:
  build: copilot/website/Dockerfile
  port: 80

cpu: 256
memory: 512
platform: linux/x86_64
count: 1

sidecars:
  proxy:
    port: 8443
    image: 123456789012.dkr.ecr.us-east-2.amazonaws.com/application-name/proxy
```

When I use the above manifest to setup an sidecar proxy, the Target Group gets setup with the original container port instead of the target container port.  Here's a look at the section of the CFT that matters:
```
      LoadBalancers:
        - ContainerName: !Ref TargetContainer <--- Here the target container is used
          ContainerPort: !Ref TargetPort <--- Here the target port is used
          TargetGroupArn: !Ref TargetGroup
      ServiceRegistries:
        - RegistryArn: !GetAtt DiscoveryService.Arn
          Port: !Ref ContainerPort <--- The original container port.  Should probably be the target port.
  TargetGroup:
    Metadata:
      'aws:copilot:description': "A target group to connect the load balancer to your service"
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      HealthCheckPath: / # Default is '/'.
      Port: !Ref ContainerPort <--- This should be the target port
      Protocol: HTTP
```

### PR DESCRIPTION
**Pull request by dannyrandall**
Fixes #4041 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/template/templates/workloads/partials/cf/alb.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ TargetGroup:
     {{- if .HTTPTargetContainer.IsHTTPS }}
     HealthCheckProtocol: HTTPS
     {{- end}}
-    Port: !Ref ContainerPort
+    Port: !Ref TargetPort
```
Comment (efekarakus) [inline internal/pkg/template/templates/workloads/partials/cf/alb.yml:29]: Is the health check port already pointing to the target port? 

This `Port` field isn't useful 😂 https://stackoverflow.com/questions/42715647/whats-the-target-group-port-for-when-using-application-load-balancer-ec2-con/42823808#42823808 but it's good to be consistent

**Review by efekarakus — COMMENTED** [review]
Looks great, just a question

File: internal/pkg/template/templates/workloads/services/backend/cf.yml [hunk 2]
```diff
@@ -102,7 +102,7 @@ Resources:
       {{- end}}
     Properties:
       {{- "\n"}}{{ include "service-base-properties" . | indent 6 }}
-      ServiceRegistries: !If [ExposePort, [{RegistryArn: !GetAtt DiscoveryService.Arn, Port: !Ref ContainerPort}], !Ref "AWS::NoValue"]
+      ServiceRegistries: !If [ExposePort, [{RegistryArn: !GetAtt DiscoveryService.Arn, Port: !Ref TargetPort}], !Ref "AWS::NoValue"]
```
Comment (iamhopaul123) [inline internal/pkg/template/templates/workloads/services/backend/cf.yml:105]: Actually this is not working. `ExposePort` is for main container, and we should check targetPort instead.

File: internal/pkg/template/templates/workloads/partials/cf/alb.yml (hunk 1 above)
Comment (dannyrandall) [reply internal/pkg/template/templates/workloads/partials/cf/alb.yml:29]: You're right, it's not useful😂 But! This field does show up in the target group console so this should make it more obvious which port we are using.

Comment (dannyrandall) [reply internal/pkg/template/templates/workloads/partials/cf/alb.yml:29]: also to answer your first question - health check port defaults to traffic port if not set manually.

File: internal/pkg/template/templates/workloads/services/backend/cf.yml (hunk 2 above)
Comment (dannyrandall) [reply internal/pkg/template/templates/workloads/services/backend/cf.yml:105]: good catch. changed backend service params to always include target port, and changed the `ExposePort` condition to use target port.

File: internal/pkg/deploy/cloudformation/stack/backend_svc.go [hunk 3]
```diff
@@ -241,18 +241,18 @@ func (s *BackendService) Parameters() ([]*cloudformation.Parameter, error) {
...
 			ParameterValue: aws.String(s.rc.EnvFileARN),
 		},
+		{
+			ParameterKey:   aws.String(WorkloadTargetContainerParamKey),
+			ParameterValue: targetContainer,
+		},
+		{
+			ParameterKey:   aws.String(WorkloadTargetPortParamKey),
+			ParameterValue: targetPort,
+		},
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/backend_svc.go:251]: 🤝

File: internal/pkg/template/workload.go [hunk 4]
```diff
@@ -235,6 +235,10 @@ type HTTPHealthCheckOpts struct {
 	DeregistrationDelay *int64
 }
 
+func (h HTTPHealthCheckOpts) IsHTTPS() bool {
```
Comment (iamhopaul123) [inline internal/pkg/template/workload.go:240]: Comment pls

File: internal/pkg/template/templates/workloads/partials/cf/alb.yml [hunk 5]
```diff
@@ -23,10 +23,14 @@ TargetGroup:
     {{- if .HTTPHealthCheck.Timeout}}
     HealthCheckTimeoutSeconds: {{.HTTPHealthCheck.Timeout}}
     {{- end}}
-    {{- if .HTTPTargetContainer.IsHTTPS }}
+    {{/* if health check port is configured to 443, or if health check port not specified and target container is https */}}
+    {{- if or (and .HTTPHealthCheck.Port .HTTPHealthCheck.IsHTTPS) (and (not .HTTPHealthCheck.Port) .HTTPTargetContainer.IsHTTPS)}}
```
Comment (iamhopaul123) [inline internal/pkg/template/templates/workloads/partials/cf/alb.yml]: Would it be better (more readable) if we move this logic to the function itself 🤔 Like can we combine

```
	HTTPTargetContainer     HTTPTargetContainer
	HTTPHealthCheck         HTTPHealthCheckOpts
```

to 

```
type HTTPTarget struct {
    HTTPTargetContainer
    HTTPHealthCheckOpts
}
```

Comment (efekarakus) [reply internal/pkg/template/templates/workloads/partials/cf/alb.yml]: 👍 or we can add a template function too, something like:
```yaml
HealthCheckProtocol: {{healthCheckProtocol .HTTPHealthCheck .HTTPTargetContainer}}
```
That returns either "HTTP" or "HTTPS"

File: internal/pkg/template/workload.go [hunk 6]
```diff
@@ -593,6 +599,24 @@ type WorkloadOpts struct {
...
+// HealthCheckProtocol returns the protocol for the Load Balancer health check,
+// or an empty string if it shouldn't be configured, defaulting to the
+// target protocol. (which is what happens, even if it isn't documented as such :))
+// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html#cfn-elasticloadbalancingv2-targetgroup-healthcheckprotocol
+func (w WorkloadOpts) HealthCheckProtocol() string {
+	switch {
+	case w.HTTPHealthCheck.Port == "443":
+		return "HTTPS"
+	case w.HTTPTargetContainer.IsHTTPS() && w.HTTPHealthCheck.Port == "":
+		return "HTTPS"
```
Comment (efekarakus) [inline internal/pkg/template/workload.go:611]: do we need this case then, or can it be ""? since the TG protocol will be set to HTTPS

**Review by efekarakus — APPROVED** [review]

File: internal/pkg/template/workload.go (hunk 6 above)
Comment (dannyrandall) [reply internal/pkg/template/workload.go:611]: i think we should explicitly configure it since the behavior isn't documented correctly

**Pull request merged**